assets/js/googlesitekit/data/utils.test.js

```diff
@@ -124,6 +128,231 @@ describe( 'data utils', () => {
+		} );
+
+		it( 'should modify combined state', () => {
+			const combinedStore = combineStores( storeOne, storeTwo );
+			let state = combinedStore.reducer();
+
+			expect( state ).toEqual( { one: 1, two: 2 } );
+
+			// It should still respond to the original actions.
+			state = combinedStore.reducer( state, combinedStore.actions.fakeActionOne() );
+			expect( state ).toEqual( { one: true, two: 2 } );
+
+			state = combinedStore.reducer( state, combinedStore.actions.fakeActionTwo() );
+			expect( state ).toEqual( { one: true, two: 2 } );
+		} );
+
+		it( 'should combine stores missing any of the keys', () => {
+			const combinedStoreMissing = {
+				INITIAL_STATE: collect( storeOne.INITIAL_STATE, storeTwo.INITIAL_STATE, storeThree.INITIAL_STATE ),
+				actions: collectActions( storeOne.actions, storeTwo.actions ),
+				controls: collectControls( storeOne.controls, storeTwo.controls ),
+				reducer: collectReducers( storeOne.reducer, storeTwo.reducer ),
+				resolvers: collectResolvers( storeOne.resolvers, storeTwo.resolvers ),
+				selectors: collectSelectors( storeOne.selectors, storeTwo.selectors ),
+			};
+			expect( JSON.stringify( combineStores( storeOne, storeTwo, storeThree ) ) ).toBe( JSON.stringify( combinedStoreMissing ) );
+
+			const combinedStoreMissingTwo = {
+				INITIAL_STATE: collect( storeOne.INITIAL_STATE, storeTwo.INITIAL_STATE ),
+				actions: collectActions( storeOne.actions, storeTwo.actions ),
+				controls: collectControls( storeOne.controls, storeTwo.controls ),
+				reducer: collectReducers( storeOne.reducer, storeTwo.reducer, storeFour.reducer ),
+				resolvers: collectResolvers( storeOne.resolvers, storeTwo.resolvers ),
+				selectors: collectSelectors( storeOne.selectors, storeTwo.selectors ),
+			};
+			expect( JSON.stringify( combineStores( storeOne, storeTwo, storeFour ) ) ).toBe( JSON.stringify( combinedStoreMissingTwo ) );
+		} );
+
+		it( 'INITIAL_STATEs, reducers, and actions should all work together even if provided by separate incomplete stores', () => {
+			const combinedStore = combineStores(
+				{ INITIAL_STATE: { one: 1 } },
+				{ reducer: combineStoresFakeReducerOne },
+				storeTwo,
+				storeFive
+			);
+			let state = combinedStore.reducer();
+
+			expect( state ).toEqual( { one: 1, two: 2 } );
+
+			// Reducer from one incomplete store responds to an action added by a different incomplete store.
+			state = combinedStore.reducer( state, combinedStore.actions.fakeActionOne() );
+			expect( state ).toEqual( { one: true, two: 2 } );
+		} );
+
+		it( 'should error if action keys are duplicated', () => {
+			expect( () => {
+				combineStores( storeOne, storeFive );
+			} ).toThrow( /collect\(\) cannot accept collections with duplicate keys. Your call to collect\(\) contains the following duplicated functions: fakeActionOne./ );
+		} );
+
+		it( 'should error if control keys are duplicated', () => {
```

Comment:
Similar to the duplicate checks for everything, I think it'd be good to have a test for each key being entirely omitted, eg:

```js
expect( () => {
  combineStores( {
    INITIAL_STATE: {},
    actions: {},
    controls: {},
    reducer: {},
    resolvers: {},
    selectors: undefined,
  } );
} ).not.toThrow();
```

for each of:

* `INITIAL_STATE`
* `actions`
* `controls`
* `reducer`
* `resolvers`
* `selectors`

Reply:
👍